### ISSUE TITLE
Zero-inflation probabilities?

### ISSUE THREAD

**Issue by kbchoi** [YosefLab/scVI#502]
Dear developers,

Thank you for a nice software package. Not really an issue but ... I am trying to find a way to get zero-inflation probability for each gene from scVI or AutoZI object. I would greatly appreciate if you give me a pointer to it. Thanks in advance.

**Comment by adamgayoso** [comment]
Hello,

You should be able to use the `generate_parameters()` method for a posterior object.

https://github.com/YosefLab/scVI/blob/6ab6aec7c031888ed535a53bc824be313416cbe0/scvi/inference/posterior.py#L696

So you'd want to create a posterior object and call this method.

**Comment by oscarclivio** [comment]
Hi @kbchoi,

First, by zero-inflation probability, do you mean :
1) either the zero-inflation component in the ZINB distribution, i.e. the probability that the cell-gene entry (it is defined per gene and cell) is zeroed out given that zero-inflation applies ?
2) or the gene is zero-inflated(as in AutoZI) ?

If 1) applies and you use scVI, you may follow directions provided by @adamgayoso.

If 2) applies, then the probability that gene g is zero-inflated is random : it is equal to 1 - delta_g where delta_g is introduced in the [preprint](https://www.biorxiv.org/content/10.1101/794875v2). As it is a random variable, you can't use it directly but you may compute its average or q(delta_g < 0.5) which can be understood as the probability that the gene has more chances to be zero-inflated than not zero-inflated.

You can draw on the [AutoZI tutorial notebook](https://github.com/YosefLab/scVI/blob/master/tests/notebooks/AutoZI_tutorial.ipynb). Specifically, you will find this probability computed for all genes in the line : 
`zi_probs = beta.cdf(0.5,alpha_posterior,beta_posterior)`

You may compute the average using : 
`zi_averages = 1. - alpha_posterior / (alpha_posterior + beta_posterior)`

If 1) applies and you use AutoZI, then it is a bit trickier... The same directions given by @adamgayoso  will give you the probability that the cell-gene entry is zeroed out given that zero-inflation applies to the gene which means, if we use the notations from the paper, m_g = 0. If you want the probability that the cell-gene entry is zeroed out without the assumption zero-inflation applies to the gene in the first place, you may integrate on m_g and it simplifies to computing the probability that the gene is zeroed out given that zero-inflation applies to the gene and multiplying it by the probability that the gene is zero-inflated which is random... So the quantity you are looking for is random. You may take its average, and this is done by computing :

`
dropouts_given_zi_applies = posterior.generate_parameters()[0]`
zi_averages = 1. - alpha_posterior / (alpha_posterior + beta_posterior)
dropouts = dropouts_given_zi_applies * zi_averages
`

I hope all of this is understandable and that you found what you were looking for !




Th

**Comment by adamgayoso** [comment]
One more follow-up, you may need to pass these values from `generate_parameters()` through the logistic function so they are interpreted as probabilities.

**Comment by oscarclivio** [comment]
> One more follow-up, you may need to pass these values from `generate_parameters()` through the logistic function so they are interpreted as probabilities.

Indeed, I had not addressed this in my answer. I will correct that

**Comment by kbchoi-jax** [comment]
Many thanks, Adam and Oscar. Sorry my question was not clear, I meant the case 1. I was able to get dropout_probs following Oscar's instruction. Running scVI just to make sure if they are giving me the same dropout probabilities. I have one more quick question:

Q: I see "posterior" has three items, and the first is related to zero-inflation (dropout) probability. Can I assume the other two quantities are shape and dispersion (or reciprocal_dispersion) parameters of negative binomial component?

**Comment by adamgayoso** [comment]
The first return item of this function is the ZI prob, the second is the mean of the NB, the third is the dispersion. Also it's recommended you call the function as `posterior.sequential().generate_parameters()` so that the outputs are in the same order as the indices used to create the posterior object.

**Comment by kbchoi-jax** [comment]
Thanks. Sorry, one more quick question.

I did the following after training scVI VAE, but the zero-inflation probabilities are different from those derived from Oscar's autoZI VAE. Is that something expected? I am guessing that's happening because zero inflation in scVI is *not* conditioned on Beta(0.5, 0.5) prior?
```
posterior = Posterior(scvi_vae, gene_dataset)
dropout_logits = posterior.sequential().generate_parameters()[0]
dropout_probs = 1. / (1. + np.exp(-dropout_logits))
```
![dropout_probs](https://user-images.githubusercontent.com/7860891/68158636-0507a680-ff1e-11e9-8840-9216b1b9b195.png)

**Comment by adamgayoso** [comment]
I think @oscarclivio will have more insight, but recall that AutoZI determines if a gene is ZI on the *gene-level* by default, when indeed there could be cell-type specific effects. Furthermore, the decision to call a gene ZI or not is equivalent to deciding x_{ng} | z_n ~ NB *or* x_{ng} | z_n ~ZINB with minimal ZI prob of 0.01 by default. 

Just to clarify, also, for the AutoZI, did you do the rescaling that @oscarclivio suggests?

**Comment by oscarclivio** [comment]
Thanks @adamgayoso for your answer!

@kbchoi : 
Indeed, for now, we do not specifically expect ZI probabilities to be equal between scVI and AutoZI. However, we did not focus on this point much when developing AutoZI and making ZI probs be similar across these two models could be interesting for further developments.

Something interesting in your plot is that we have a large number of genes with near-zero ZI prob on scVI but much higher ZI prob on AutoZI. The 0.01 threshold is not enough to explain this. It is possible that AutoZI chooses the NB model (e.g. delta_g near 1 on average) but the ZI prob given ZI happening tries to compensate this and does it so much that we end up with high ZI probs. But this is only an assumption.
Also, the `UnsupervisedTrainer` class trains AutoZI and scVI differently : by default, the optional parameter `normalize_loss` is set at False for AutoZI but True for scVI. Hence, the differences in training may result in differences in ZI probs.
How is your plot if you enter `normalize_loss=True` as an input of the `UnsupervisedTrainer` instance you use for both AutoZI and scVI ? For `normalize_loss=False` ? 

If you would like to share your piece of code or notebook more thoroughly but would like to keep it private, please feel free to contact me at `oclivio {AT} live {dot} fr` !

**Comment by kbchoi** [comment]
Thank you for the response @adamgayoso and @oscarclivio. I used the following ipython notebook tutorials on the github, both of them use `UnsupervisedTrainer` from `scvi.inference`. Therefore, I found both are using `normalized_loss=None`.

[AutoZI tutorial](https://nbviewer.jupyter.org/github/YosefLab/scVI/blob/master/tests/notebooks/AutoZI_tutorial.ipynb)
[scVI basic tutorial](https://nbviewer.jupyter.org/github/YosefLab/scVI/blob/master/tests/notebooks/basic_tutorial.ipynb)

When I set `normalized_loss=True` for both, the scatter plot is like this.

![scvi](https://user-images.githubusercontent.com/47792282/68297982-473bff80-0066-11ea-9a48-28a5ec6d9182.png)

I put my data, a pickled `GeneExpressionDataset`  of `scvi.dataset` on my ftp site (ftp://churchill-lab.jax.org/analysis/AutoZI/mmHeart.pkl) in case you want to take a look by yourself.

**Comment by adamgayoso** [comment]
Closing due to inactivity.
